safemysql.class.php

```diff
@@ -514,38 +555,60 @@ public function getStats()
 		return $this->stats;
 	}
 
-	/**
-	 * private function which actually executes $query.
-	 * also logs some stats like profiling info and error message
-	 * 
-	 * @param string $query - a regular SQL query
-	 * @return mysqli result resource or FALSE on error
-	 */
+
 	private function rawQuery($query)
 	{
-		$start = microtime(TRUE);
-		$res   = mysqli_query($this->conn, $query);
-		$timer = microtime(TRUE) - $start;
-
-		$this->stats[] = array(
-			'query' => $query,
-			'start' => $start,
-			'timer' => $timer,
-		);
-		if (!$res)
-		{
-			$error = mysqli_error($this->conn);
-			$errno = mysqli_errno($this->conn);
-			
-			end($this->stats);
-			$key = key($this->stats);
-			$this->stats[$key]['error'] = $error;
+		$executeQuery = function ($db) use ($query) {
+			$start = microtime(TRUE);
+			$res   = mysqli_query($this->conn, $query);
+			$timer = microtime(TRUE) - $start;
+
+			$this->stats[] = array(
+				'query' => $query,
+				'start' => $start,
+				'timer' => $timer,
+			);
+			if (!$res)
+			{
+				$error = mysqli_error($this->conn);
+				$errno = mysqli_errno($this->conn);
+
+				end($this->stats);
+				$key = key($this->stats);
+				$this->stats[$key]['error'] = $error;
+				$this->cutStats();
+
+				$this->error("$error. Full query: [$query]", $errno);
+			}
 			$this->cutStats();
-			
-			$this->error("$error. Full query: [$query]", $errno);
+			return $res;
+		};
+
+		if ($this->retryOnDeadlock && !empty($this->transactionInProgress)) {
```

Comment:
- [ ] Waaaaaait, the second condition here is the wrong way round, surely? I think this justifies my remark about the use of `empty()` being needlessly confusing above.